src/sage/rings/polynomial/polynomial_element.pyx

```diff
@@ -2168,162 +2168,151 @@ cdef class Polynomial(CommutativePolynomial):
-                    if T == 0:
-                        continue
+                return -f[0] / f[1]
+            return ring(-f[0] / f[1])
+
+        def _any_root_squarefree(f, ring=None):
+            # Compute the distinct degree factorisation for degree one, 
+            # the output will be the product of #all degree one 
+            # irreducible polynomials dividing f
+            q = f.base_ring().order() # p^k
+            R, X = f.parent().objgen()
+            w = pow(X, q, f) - X
+            g = f.gcd(w)
+
+            # When the degree is one, we have found our
+            # root. If the degree is zero, there are no roots
+            if g.degree().is_one():
+                return _linear_root(g, ring=ring)
+            elif g.degree().is_zero():
+                raise ValueError(f"no root can be computed for {f}")
+
+            # Otherwise, we can find a root from the CZ splitting of g
+            return _cantor_zassenhaus_split_root(g, 1, ring=ring)
+
+        def _cantor_zassenhaus_split_root(f, degree, ring=None):
+            R = f.parent()
+            q = f.base_ring().order()
+            # Need to handle odd and even charcteristic separately
+            if q % 2 != 0:
+                # We expect to succeed with 1/2 probability, so if we
+                # try 1000 times and fail, there's a bug somewhere.
+                for _ in range(1000):
+                    T = R.random_element(2*degree + 1)
+                    if T.is_zero(): continue # skip T = 0
                     T = T.monic()
-                    h = self.gcd(pow(T, Integer((q**degree-1)/2), self)-1)
+
+                    # Compute the gcd. 50% chance this is a non-trivial factor of f
+                    h = f.gcd(pow(T, (q-1)//2, f)-1)
                     hd = h.degree()
-                    if hd != 0 and hd != self.degree():
-                        if 2*hd <= self.degree():
-                            return h.any_root(ring, -degree, True)
+
+                    # If we found a degree-one root, return it
+                    if hd.is_one():
+                        return _linear_root(h, ring=ring) 
+
+                    # Else check if we have a non-trivial factor and keep going
+                    if not hd.is_zero() and hd != f.degree():
+                        if 2*hd <= f.degree():
+                            return _cantor_zassenhaus_split_root(h, degree, ring=ring)
                         else:
-                            return (self//h).any_root(ring, -degree, True)
-        else:
-            rs = self.roots(ring=ring, multiplicities=False)
-            if rs:
-                return rs[0]
-            raise ValueError("no roots (non-field) %s" % self)
+                            return _cantor_zassenhaus_split_root(f//h, degree, ring=ring)
+                # If you are reaching this error, chances are there's a bug in the code.
+                raise ValueError(f"no splitting of degree {degree} found for {f}")
```

Comment:
This should be an `AssertionError`: It's something that should never happen unless the code is wrong somehow.